src/modules/train_controller/src/train_controller.cc

```diff
@@ -1 +1,324 @@
+{
+    brake_failure_ = state;
+}
+
+void SoftwareTrainController::SetSignalPickupFailure(const bool state)
+{
+    signal_pickup_failure_ = state;
+}
+
+void SoftwareTrainController::SetEngineFailure(const bool state)
+{
+    engine_failure_ = state;
+}
+
+void SoftwareTrainController::SetCommandedInternalTemperature(const types::DegreesFahrenheit temp)
+{
+    commanded_internal_temperature_ = temp;
+}
+
+void SoftwareTrainController::SetActualInternalTemperature(const types::DegreesFahrenheit temp)
+{
+    actual_internal_temperature_ = temp;
+}
+
+void SoftwareTrainController::SetAuthority(const types::Meters authority)
+{
+    authority_ = authority;
+}
+
+void SoftwareTrainController::SetKP(const uint16_t kp)
+{
+    kp_ = kp;
+}
+void SoftwareTrainController::setKI(const uint16_t ki)
+{
+    ki_ = ki;
+}
+
+void SoftwareTrainController::SetArrived(const bool arrived)
+{
+    arrived_ = arrived;
+}
+
+void SoftwareTrainController::CalculateCommandedPower()
+{
+    // P(t) = Kp*[V_cmd(t) - v(t)]  +  Ki*∫[Vcmd(τ) - ActualSpeed(τ)]dτ
+    // A function in time that represents the PI Controller
+
+    float blockSpeedLimit = 50.0; // default speed in Km/H
+
+    // Defining Vcmd and Actual speed in m/s
+    types::MetersPerSecond Vsetpoint = driver_speed_;
+
+    // convert to m/s from km/hr
+    blockSpeedLimit = convert::KilometersPerHourToMetersPerSecond(blockSpeedLimit);
+
+    if (Vsetpoint > blockSpeedLimit)
+    {
+        Vsetpoint = blockSpeedLimit;
+    }
+
+    // Calculating Verror
+    double Verror = Vsetpoint - current_speed_;
+
+    // Calculating Kp term
+    float KPterm = Verror * kp_;
+
+    // Temp time passed since last update
+    float deltaTime = 1;
+
+    // This section is where the integral section of the equation will be calculated
+    integral_sum_ += Verror * deltaTime;
+
+    // Calculating Ki term
+    float KIterm = ki_ * integral_sum_;
+
+
+    if (emergency_brake_ == true || arrived_)
+    {
+        integral_sum_    = 0;
+        commanded_power_ = 0;
+    }
+    else
+    {
+        //Checking if Current Train Velocity is greater than Setpoint speed
+        if (current_speed_ > driver_speed_)
+        {
+            integral_sum_    = 0;
+            commanded_power_ = 0;
+
+            double speed_difference = current_speed_ - driver_speed_;
+
+            //Function to to assign service brake
+            CalculateServiceBrake(speed_difference);
+        }
+        //Checking if Service brake is on
+        else
+        if (service_brake_percentage_ > 0)
+        {
+            integral_sum_    = 0;
+            commanded_power_ = 0;
+        }
+        //Normal power calculation
+        else
+        {
+
+            commanded_power_ = KPterm + KIterm;
+
+            if (commanded_power_ > max_power_)
+            {
+                commanded_power_ = max_power_;
+            }
+        }
+
+
+    }
+}
+
+void SoftwareTrainController::CalculateServiceBrake(double speed_difference)
```

Comment:
I think this whole function can be condensed to a single calculation, but I want to make sure I'm understanding the intention